### PR TITLE
[flang] Lowering changes for assigning dummy_scope to hlfir.declare.

### PR DESCRIPTION
**Pull request by vzakhari**
The lowering produces fir.dummy_scope operation if the current
function has dummy arguments. Each hlfir.declare generated
for a dummy argument is then using the result of fir.dummy_scope
as its dummy_scope operand. This is only done for HLFIR.

I was not able to find a reliable way to identify dummy symbols
in `genDeclareSymbol`, so I added a set of registered dummy symbols
that is alive during the variables instantiation for the current
function. The set is initialized during the mapping of the dummy
argument symbols to their MLIR values. It is reset right after
all variables are instantiated - this is done to avoid generating
hlfir.declare operations with dummy_scope for the clones of
the dummy symbols (e.g. this happens with OpenMP privatization).

If this can be done in a cleaner way, please advise.

### CODE REVIEW

**Review by tblah — APPROVED** [review]
LGTM. I can't think of a cleaner way to do this off the top of my head. You could fix all of the OpenMP code, but it is very likely that more will be added so I think the way you have done it is a better option.

**Review by jeanPerier — APPROVED** [review]
Thanks you Slava! You approach seems safer. Without more context in genDeclareSymbol, you cannot make a difference between internal procedure dummy vs host procedure dummy symbols for instance.

Good job updating all those tests, did you manage scripting it?

File: flang/lib/Lower/Bridge.cpp [hunk 1]
```diff
@@ -5375,6 +5400,23 @@ class FirConverter : public Fortran::lower::AbstractConverter {
...
+  /// It will be used to set dummy_scope operand of the hlfir.declare
+  /// operations.
+  void setDummyArgsScope(mlir::Value val) {
+    assert(!dummyArgsScope && val);
+    dummyArgsScope = val;
+  }
+
+  /// Record the given symbol as a dummy argument of this function.
+  void registerDummySymbol(Fortran::semantics::SymbolRef symRef) {
+    auto *sym = symRef->HasLocalLocality() ? &*symRef : &symRef->GetUltimate();
```
[the remark this answers is not in the document]
Comment (vzakhari) [reply flang/lib/Lower/Bridge.cpp:5413]: Nope, I do not have an example for this.  I was just following the `lookupSymbol` impl.  I will remove it here and in `isRegisteredDummySymbol`. Thanks!

**Comment by vzakhari** [comment]
> Thanks you Slava! You approach seems safer. Without more context in genDeclareSymbol, you cannot make a difference between internal procedure dummy vs host procedure dummy symbols for instance.
> 
> Good job updating all those tests, did you manage scripting it?

I did not find a reliable way to script this, so I added `dummy_scope ...` to the declare ops, and then used this as an "anchor" to add `!fir.dscope` into the type signature of the operation.

**Pull request merged**